### PR TITLE
Merge Testing/#0001 android into main 

### PR DESCRIPTION
**Pull request by AnnaSolvd**

### CODE REVIEW

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@DeviceType(pageType = DeviceType.Type.IOS_PHONE, parentClass = CheckOutInformationIOSBase.class)
+public class CheckOutInformationIOS extends CheckOutInformationIOSBase {
+
+    private static final Logger logger = LoggerFactory.getLogger(CheckOutInformationIOS.class);
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeTextField[`name == 'test-First Name'`]")
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java:18]: Can be replaced with just Predicate

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private ExtendedWebElement firstNameInput;
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeTextField[`name == 'test-Last Name'`]")
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java:21]: Can be replaced with just Predicate

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private ExtendedWebElement lastNameInput;
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeTextField[`name == 'test-Zip/Postal Code'`]")
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java:24]: Can be replaced with just Predicate

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    private ExtendedWebElement zipCodeInput;
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeOther[`name == 'test-CONTINUE'`]")
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/CheckOutInformationIOS.java:27]: Can be replaced with just Predicate

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    private SwagLabsProduct selectedProduct;
+
+    public HomePageIOS(WebDriver driver) {
+        super(driver);
+        setUiLoadedMarker(cartButton);
+        logger.info("HomePageIOS open");
+        selectedProduct = selectProduct();
+    }
+
+    private SwagLabsProduct selectProduct() {
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:48]: Shouldn't be selectRandomProduct/getRandomProduct?

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public LeftNavigationSidebar clickMenuButton() {
+        clickButton(menuButton);
+        logger.info("Click menu button");
+        return new LeftNavigationSidebar(getDriver());
+    }
+
+    @Override
+    public boolean checkFirstProductPrice(String price) {
+        String productPrice = productList.get(0).getPrice();
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:93]: I think you can replace get(0) with getFirst()

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        String cleanedPrice = price.replaceAll("[^\\d.]", "");
+        logger.info("Product price: {}", cleanedPrice);
+        return productPrice.contains(price);
+    }
+
+    @Override
+    public SwagLabsProduct getRandomProduct() {
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:100]: Why do you have two methods do same thing selectProduct and getRandomProduct?

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/LoginPageIOS.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@DeviceType(pageType = DeviceType.Type.IOS_PHONE, parentClass = LoginPageIOSBase.class)
+public class LoginPageIOS extends LoginPageIOSBase {
+
+    private static final Logger logger = LoggerFactory.getLogger(LoginPageIOS.class);
+
+    @ExtendedFindBy(iosClassChain = "**/XCUIElementTypeTextField[`name == 'test-Username'`]")
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/LoginPageIOS.java:18]: isoPredicate can be used

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        return selectedProduct;
+    }
+
+    @Override
+    public ProductPageIOSBase clickRandomProduct() {
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:105]: Could delegate this method to SwagLabsProduct and from there initPage of ProductPage. So you can get random product and then from instance of the product you can just click on it and get ProductPage

**Review by MariuszSolvd — CHANGES_REQUESTED** [review]
Ok, I checked iOS overall it looks good. I made some small comments. Some of locators can be changed to iosPredicate, I prefer these ones since it's easier to read. You can change it easily. I mark some of these fields, but some I didn't, since I didn't want to spam you too much with comments. And some minor improvements that I suggested.

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 6 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:93]: I wanted to use getFirst() but got an error. It turned out that getFirs() is from JDK 21, and in the project I had JDK 14

**Review by whetman — APPROVED** [review]
Looks good to me

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 9 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java]: Done

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 7 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:90]: Done

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 6 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java]: Done

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 5 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java]: Done

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 6 above)
Comment (MariuszSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java]: Yeah, I saw that you set from java 11 to java 14 in your POM, and I thought you just like to use Java 14 that's why you use 14 xD

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/LoginPageIOS.java (hunk 8 above)
Comment (AnnaSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/LoginPageIOS.java]: I changed iosClassChain to iosPredicate everywhere I could

File: src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java (hunk 7 above)
Comment (MariuszSolvd) [reply src/main/java/com/solvd/swaglabs/gui/pages/ios/HomePageIOS.java:90]: I see you deleted selectProduct() and now use just getRandomProduct(), good job.

File: src/main/java/com/solvd/swaglabs/gui/components/SwagLabsProduct.java [hunk 10]
```diff
@@ -42,8 +44,10 @@ public void addProductToCart() {
         addToCartButton.click();
     }
 
-    public void clickProduct() {
+    public ProductPageIOSBase clickProduct() {
```
Comment (MariuszSolvd) [inline src/main/java/com/solvd/swaglabs/gui/components/SwagLabsProduct.java:47]: Ok, I can see you add click in SwagLabsProduct, I like it!

**Review by MariuszSolvd — APPROVED** [review]
Looks good, well done!

**Pull request merged**